tgui/packages/tgui/components/MenuBar.tsx

```diff
@@ -0,0 +1,231 @@
+import { logger } from '../logging';
+import { Icon } from './Icon';
+
+type MenuProps = {
+  children: any;
+  width: string;
+  menuRef: RefObject<HTMLElement>;
+  onOutsideClick: () => void;
+};
+
+class Menu extends Component<MenuProps> {
+  private readonly handleClick: (event) => void;
+
+  constructor(props) {
+    super(props);
+    this.handleClick = (event) => {
+      if (!this.props.menuRef.current) {
+        logger.log(`Menu.handleClick(): No ref`);
+        return;
+      }
+
+      if (this.props.menuRef.current.contains(event.target)) {
+        logger.log(`Menu.handleClick(): Inside`);
+      } else {
+        logger.log(`Menu.handleClick(): Outside`);
+        this.props.onOutsideClick();
+      }
+    };
+  }
+
+  // eslint-disable-next-line react/no-deprecated
+  componentWillMount() {
+    window.addEventListener('click', this.handleClick);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('click', this.handleClick);
+  }
+
+  render() {
+    const { width, children } = this.props;
+    return (
+      <div
+        className={'MenuBar__menu'}
+        style={{
+          width: width,
+        }}>
+        {children}
+      </div>
+    );
+  }
+}
+
+type MenuBarDropdownProps = {
+  open: boolean;
+  openWidth: string;
+  children: any;
+  disabled?: boolean;
+  display: InfernoNode;
+  onMouseOver: () => void;
+  onClick: () => void;
+  onOutsideClick: () => void;
+  className?: string;
+};
+
+class MenuBarButton extends Component<MenuBarDropdownProps> {
+  private readonly menuRef: RefObject<HTMLDivElement>;
+
+  constructor(props) {
+    super(props);
+    this.menuRef = createRef();
+  }
+
+  render() {
+    const { props } = this;
+    const {
+      open,
+      openWidth,
+      children,
+      disabled,
+      display,
+      onMouseOver,
+      onClick,
+      onOutsideClick,
+      ...boxProps
+    } = props;
+    const { className, ...rest } = boxProps;
+
+    return (
+      <div ref={this.menuRef}>
+        <Box
+          className={classes([
+            'MenuBar__MenuBarButton',
+            'MenuBar__font',
+            'MenuBar__hover',
+            className,
+          ])}
+          {...rest}
+          onClick={disabled ? undefined : onClick}
+          onmouseover={onMouseOver}>
+          <span className="MenuBar__MenuBarButton-text">{display}</span>
+        </Box>
+        {open && (
+          <Menu
+            width={openWidth}
+            menuRef={this.menuRef}
+            onOutsideClick={onOutsideClick}>
+            {children}
+          </Menu>
+        )}
+      </div>
+    );
+  }
+}
+
+type MenuBarItemProps = {
+  entry: string;
+  children: any;
+  openWidth: string;
+  display: InfernoNode;
+  setOpenMenuBar: (entry: string | null) => void;
+  openMenuBar: string | null;
+  setOpenOnHover: (flag: boolean) => void;
+  openOnHover: boolean;
+  disabled?: boolean;
+  className?: string;
+};
+
+export const Dropdown = (props: MenuBarItemProps) => {
```

Comment:
This component seems to be a bit useless, passes through a lot of props, has some logic for managing state yet delegates it to a parent component, smells like very tight coupling. Why not just inline everything that it does?